File: PAT_Basic_Level_by_java/1052 卖个萌/1052 卖个萌.java
/**
 * 1052 卖个萌
 * 
 * 题目描述：
 * 萌萌哒表情符号通常由手、眼、口三个基本部分组成。
 * 给定三行字符串，分别代表可用的手、眼、口符号集合（每个符号用方括号 [] 包裹）。
 * 用户给定 K 个请求，每个请求给出 5 个序号（左手、左眼、口、右眼、右手），
 * 程序按要求输出对应的表情。若序号超出范围，则输出 "Are you kidding me? @\\/@"
 * 
 * 实现原理：
 * 解题思路：
 * 1. 使用正则表达式 \\[([^\\]]+)\\] 从每行中提取方括号内的符号内容。
 * 2. 将提取出的手、眼、口符号分别存储在三个 List<String> 中。
 * 3. 对每个用户请求，检查 5 个序号是否在有效范围内（1 到对应列表大小）。
 * 4. 若合法，按 "[左手](左眼口右眼)[右手]" 的格式拼接输出；
 *    若不合法，输出预设的调侃语句。
 * 
 * 核心思想：正则解析 + 列表存储 + 下标映射。注意序号从 1 开始，需减 1 转换为下标。
 * 
 * 时间复杂度：O(N + K)，N 为符号总数，K 为用户请求数，空间复杂度：O(N)
 */
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 解析手、眼、口的符号集合
        List<String> hands = parseSymbols(br.readLine());
        List<String> eyes = parseSymbols(br.readLine());
        List<String> mouths = parseSymbols(br.readLine());

        int K = Integer.parseInt(br.readLine().trim());

        // 处理每个用户请求
        for (int i = 0; i < K; i++) {
            String[] parts = br.readLine().trim().split("\\s+");
            int leftHand = Integer.parseInt(parts[0]);
            int leftEye = Integer.parseInt(parts[1]);
            int mouth = Integer.parseInt(parts[2]);
            int rightEye = Integer.parseInt(parts[3]);
            int rightHand = Integer.parseInt(parts[4]);

            // 检查所有序号是否合法（序号从 1 开始）
            if (leftHand < 1 || leftHand > hands.size() ||
                rightHand < 1 || rightHand > hands.size() ||
                leftEye < 1 || leftEye > eyes.size() ||
                rightEye < 1 || rightEye > eyes.size() ||
                mouth < 1 || mouth > mouths.size()) {
                System.out.println("Are you kidding me? @\\/@");
            } else {
                // 按格式拼接表情：[左手](左眼口右眼)[右手]
                System.out.println(hands.get(leftHand - 1) + "(" +
                    eyes.get(leftEye - 1) + mouths.get(mouth - 1) + eyes.get(rightEye - 1) + ")" +
                    hands.get(rightHand - 1));
            }
        }
    }

    /**
     * 从字符串中解析方括号 [] 内的符号
     * @param line 包含符号的原始字符串
     * @return 符号列表
     */
    private static List<String> parseSymbols(String line) {
        List<String> symbols = new ArrayList<>();
        Pattern pattern = Pattern.compile("\\[([^\\]]+)\\]");
        Matcher matcher = pattern.matcher(line);
        while (matcher.find()) {
            symbols.add(matcher.group(1));
        }
        return symbols;
    }
}

File: PAT_Basic_Level_by_py3/1052 卖个萌/1052 卖个萌.py
# 题目：1052 卖个萌
#
# 题目描述：
#   萌萌哒表情符号通常由“手”、“眼”、“口”三个主要部分组成。简单起见，我们假设一个表情符号是按下列格式输出的：
#   ```
#   [左手]([左眼][口][右眼])[右手]
#   ```
#   现给出可选用的符号集合，请你按用户的要求输出表情。
#
# 输入格式：
#   输入首先在前三行顺序对应给出手、眼、口的可选符号集。每个符号括在一对方括号 `[]`内。题目保证每个集合都至少有一个符号，并不超过 10 个符号；每个符号包含 1 到 4 个非空字符。
#   之后一行给出一个正整数 K，为用户请求的个数。随后 K 行，每行给出一个用户的符号选择，顺序为左手、左眼、口、右眼、右手——这里只给出符号在相应集合中的序号（从 1 开始），数字间以空格分隔。
#
# 输出格式：
#   对每个用户请求，在一行中输出生成的表情。若用户选择的序号不存在，则输出 `Are you kidding me? @\/@`。
#
# 实现原理：
#   - 循环迭代处理
#   - 列表操作
#
# 算法思路：
#   字符串处理：解析表情符号，
#   根据用户输入组装输出结果。
#
import re
import sys

hands = re.findall(r'\[([^\]]*)\]', sys.stdin.readline())
eyes = re.findall(r'\[([^\]]*)\]', sys.stdin.readline())
mouths = re.findall(r'\[([^\]]*)\]', sys.stdin.readline())

for _ in range(int(sys.stdin.readline())):
    left_hand, left_eye, mouth, right_eye, right_hand = map(
        int, sys.stdin.readline().split()
    )
    if (
        1 <= left_hand <= len(hands)
        and 1 <= right_hand <= len(hands)
        and 1 <= left_eye <= len(eyes)
        and 1 <= right_eye <= len(eyes)
        and 1 <= mouth <= len(mouths)
    ):
        print(
            hands[left_hand - 1]
            + '('
            + eyes[left_eye - 1]
            + mouths[mouth - 1]
            + eyes[right_eye - 1]
            + ')'
            + hands[right_hand - 1]
        )
    else:
        print('Are you kidding me? @\\/@')
